core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -92,7 +92,7 @@ private enum State
     private final BooleanSupplier anySourceTaskBlocked;
     private final PartitionIdAllocator partitionIdAllocator;
     private final Map<InternalNode, RemoteTask> scheduledTasks;
-    private final List<Split> pendingSplits = new ArrayList<>();
```

Comment:
see the line below
```
splitAssignment.values().forEach(pendingSplits::remove); // AbstractSet.removeAll performs terribly here.
```

when pendingSplits is a list, this will be `List.remove`, resulting in O(n^2) per pending split count (connector batch size)